demo/c/porcupine_demo_mic.c

```diff
@@ -9,23 +9,154 @@
     specific language governing permissions and limitations under the License.
 */
 
-#include <alsa/asoundlib.h>
+#define MINIAUDIO_IMPLEMENTATION
+#include "miniaudio/miniaudio.h"
+
+#if defined(WIN32) || defined(WIN64)
+#include <windows.h>
+#else
 #include <dlfcn.h>
+#endif
+
 #include <signal.h>
 #include <stdio.h>
 
 #include "pv_porcupine.h"
 
+typedef struct frame_buffer {
+    int16_t *buffer;
+    int32_t max_size;
+    int32_t filled;
+} frame_buffer_t;
+
+typedef struct pv_porcupine_data {
+    frame_buffer_t  frame_buffer;
+    const char      *(*pv_status_to_string_func)(pv_status_t);
+    pv_status_t     (*pv_porcupine_process_func)(pv_porcupine_t *, const int16_t *, int32_t *);
+    pv_porcupine_t  *porcupine;
+} pv_porcupine_data_t;
+
 static volatile bool is_interrupted = false;
 
+void *open_dl(const char *dl_path) {
+#if defined(WIN32) || defined(WIN64)
+    return LoadLibrary(dl_path);
+#else
+    return dlopen(dl_path, RTLD_NOW);
+#endif
+}
+
+void *load_symbol(void *handle, const char *symbol) {
+#if defined(WIN32) || defined(WIN64)
+    return GetProcAddress((HMODULE) handle, symbol);
+#else
+    return dlsym(handle, symbol);
+#endif
+}
+
+void close_dl(void *handle) {
+#if defined(WIN32) || defined(WIN64)
+    FreeLibrary((HMODULE) handle);
+#else
+    dlclose(handle);
+#endif
+}
+
+void print_dl_error(const char *message) {
+#if defined(WIN32) || defined(WIN64)
+    fprintf(stderr, "%s with code '%lu'.\n", message, GetLastError());
+#else
+    fprintf(stderr, "%s with '%s'.\n", message, dlerror());
+#endif
+}
+
 void interrupt_handler(int _) {
     (void) _;
     is_interrupted = true;
 }
 
+void data_callback_fixed(ma_device* device, void* output, const void* input, ma_uint32 frame_count){
```

Comment:
does this function need to have the same interface as `data_callback`? as far as I understand, it can be some arbitrary function, right?
So if I am correct, let's change it to something like:
`void porcupine_process_callback(pv_porcupine_data_t *pv_porcupine_data, int16_t *pcm)`

the current name/interface is a little bit confusing. it seems like a function related to audio/mic, but it is actually the heart of the program. Also we want to show that the main function related to porcupine is pretty simple and short. The other parts are related to setup the environment/audio stream.